Libplanet/Store/TrieStateStore.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Collections.Immutable;
+using System.Security.Cryptography;
+using System.Text;
+using Bencodex.Types;
+using Libplanet.Action;
+using Libplanet.Blocks;
+using Libplanet.Store.Trie;
+using Libplanet.Store.Trie.Nodes;
+
+namespace Libplanet.Store
+{
+    /// <summary>
+    /// An <see cref="IStateStore"/> implementation. It stores states with <see cref="Trie"/>.
+    /// </summary>
+    public class TrieStateStore : IStateStore
+    {
+        private readonly IKeyValueStore _stateKeyValueStore;
+        private readonly IKeyValueStore _stateHashKeyValueStore;
+
+        /// <summary>
+        /// Creates a new <see cref="TrieStateStore"/>.
+        /// </summary>
+        /// <param name="stateKeyValueStore">The storage to store states. It used by
+        /// <see cref="Trie"/> in internal.</param>
+        /// <param name="stateHashKeyValueStore">The storage to store state hash corresponding to
+        /// block hash.</param>
+        public TrieStateStore(
+            IKeyValueStore stateKeyValueStore, IKeyValueStore stateHashKeyValueStore)
+        {
+            _stateKeyValueStore = stateKeyValueStore;
+            _stateHashKeyValueStore = stateHashKeyValueStore;
+        }
+
+        /// <inheritdoc/>
+        public void SetStates<T>(
+            HashDigest<SHA256> blockHash,
+            IImmutableDictionary<string, IValue> states,
+            Func<HashDigest<SHA256>, Block<T>> blockGetter)
+            where T : IAction, new()
+        {
+            var block = blockGetter(blockHash);
+            var previousBlockStateHashBytes = block?.PreviousHash is null
+                ? null
+                : _stateHashKeyValueStore.Get(block.PreviousHash.Value.ToByteArray());
+            var trieRoot = previousBlockStateHashBytes is null
+                ? null
+                : new HashNode(new HashDigest<SHA256>(previousBlockStateHashBytes));
+
+            var prevStatesTrie = new Trie.Trie(_stateKeyValueStore, trieRoot);
```

Comment:
It can be a signal that `Libplanet.Store.Trie` isn't good name not enough. 🤔  (see also. https://docs.microsoft.com/ko-kr/dotnet/standard/design-guidelines/names-of-namespaces#namespaces-and-type-name-conflicts)

Reply:
Yes, you're right but I don't have a cool idea. Is `MerkleTrie` okay as a good name?

Reply:
`MerkleTrie` seems good for type name.